### ISSUE TITLE
Speed up `peek` for ordered containers

### ISSUE THREAD

**Issue by sjakobi** [fpco/store#95]
As far as I can tell, the `peek`s for `Set`, `IntSet`, `Map` and `IntMap` are currently implemented via the order-agnostic `setFromList` and `mapFromList` functions. Given that all these data structures are `poke`d in order, they could likely be `peek`ed slightly faster in terms of `fromDistinctAscList`.

---
BTW, I've always wondered whether in the `peek` implementation for e.g. `Set`, i.e. `setFromList <$> peek`, the intermediary list is fused away, or if the entire list has to be constructed in order to check that the monadic operation was indeed successful, and only then is the set constructed from the list.
If the list is fully constructed, I think a faster "streaming" (?) implementation could be found.

(Note to myself: It might be possible to test this with [list-fusion-probe](http://hackage.haskell.org/package/list-fusion-probe).)

I'd be happy if anyone felt like explaining this to me, or would like to investigate this together! :)

**Comment by sjakobi** [comment]
It also probably couldn't hurt to add a few benchmarks for these data structures (and other Store instances too)…

**Comment by sjakobi** [comment]
Regarding list fusion in `peek`: I've tried replacing the definition of `peekMap` with `(mapFromList . fuseThis) <$> peek` which resulted in an error, indicating that the list is not fused away.

I also looked at the deserialization implementations in `binary` and `cereal` which to my surprise use the same simple `fmap fromList` scheme as `store`. (An alternative explanation is of course that this is the right way (TM) to deserialize maps and sets)

I guess I should set up some benchmarks with weigh and criterion and then maybe look first at the `Store` instance for `HashSet` which has [a particularly simple `fromList` implementation](http://hackage.haskell.org/package/unordered-containers-0.2.7.2/docs/src/Data-HashSet.html#fromList).
